src/devices/Bmxx80/Bme280.cs

```diff
@@ -81,30 +85,54 @@ public Sampling HumiditySampling
         /// Contains <see cref="double.NaN"/> otherwise.
         /// </param>
         /// <returns><code>true</code> if measurement was not skipped, otherwise <code>false</code>.</returns>
-        public bool TryReadHumidity(out double humidity)
+        [Obsolete("TryReadHumidity is deprecated, please use Read or ReadAsync instead.")]
+        public bool TryReadHumidity(out double humidity) => TryReadHumidityInternal(out humidity);
+
+        /// <summary>
+        /// Gets the required time in ms to perform a measurement with the current sampling modes.
+        /// </summary>
+        /// <returns>The time it takes for the chip to read data in milliseconds rounded up.</returns>
+        public override int GetMeasurementDuration()
         {
-            if (HumiditySampling == Sampling.Skipped)
+            return s_osToMeasCycles[(int)PressureSampling] + s_osToMeasCycles[(int)TemperatureSampling] + s_osToMeasCycles[(int)HumiditySampling];
+        }
+
+        /// <summary>
+        /// Performs a synchronous reading.
+        /// </summary>
+        /// <returns><see cref="Bme280ReadResult"/></returns>
+        public Bme280ReadResult Read()
+        {
+            if (ReadPowerMode() != Bmx280PowerMode.Normal)
             {
-                humidity = double.NaN;
-                return false;
+                SetPowerMode(Bmx280PowerMode.Forced);
+                Thread.Sleep(GetMeasurementDuration());
             }
 
-            // Read the temperature first to load the t_fine value for compensation.
-            TryReadTemperature(out _);
-
-            var hum = Read16BitsFromRegister((byte)Bme280Register.HUMIDDATA, Endianness.BigEndian);
+            TryReadTemperatureInternal(out var temperature);
+            TryReadPressureInternal(out var pressure, true);
+            TryReadHumidityInternal(out var humidity, true);
 
-            humidity = CompensateHumidity(hum);
-            return true;
+            return new Bme280ReadResult { Temperature = temperature, Pressure = pressure, Humidity = humidity };
         }
 
         /// <summary>
-        /// Gets the required time in ms to perform a measurement with the current sampling modes.
+        /// Performs an asynchronous reading.
         /// </summary>
-        /// <returns>The time it takes for the chip to read data in milliseconds rounded up.</returns>
-        public override int GetMeasurementDuration()
+        /// <returns><see cref="Bme280ReadResult"/></returns>
+        public async Task<Bme280ReadResult> ReadAsync()
         {
-            return s_osToMeasCycles[(int)PressureSampling] + s_osToMeasCycles[(int)TemperatureSampling] + s_osToMeasCycles[(int)HumiditySampling];
+            if (ReadPowerMode() != Bmx280PowerMode.Normal)
+            {
+                SetPowerMode(Bmx280PowerMode.Forced);
+                await Task.Delay(GetMeasurementDuration());
+            }
+
+            TryReadTemperatureInternal(out var temperature);
```

Comment:
should you return false if any of these fails?

Reply:
That's actually an open question I had. See comment https://github.com/dotnet/iot/pull/797#issuecomment-632354524. In the past if one of the measurements was skipped by the user the value would be NaN. Since UnitsNet doesn't handle NaN we need another solution (null or max value).

Reply:
If you use `default(Temperature)` as result then you can check i.e. `temp.Unit == TemperatureUnit.Undefined`

Reply:
the old methods returning double can check for that and return NaN instead

Reply:
Using `default` would result in:

![Capture](https://user-images.githubusercontent.com/27669180/83336240-05003700-a2b2-11ea-8fb9-5a62ae21f143.PNG)

(Temperature + Humidity + Pressure measurement skipped)

Reply:
I'm using

```csharp
            Temperature temp = default;
            Console.WriteLine(temp);
            Console.WriteLine(temp.Unit);
            Console.WriteLine(temp.DegreesCelsius);
```

and it prints:

```
0 K
Kelvin
-273.15
```

unfortunately it's not throwing any errors or return bogus units but you can tell immediately the value is wrong.

For Try* methods user is responsible for checking the output (similarly as if you used Dictionary with Value being int you'd get 0). For existing methods if possible we should hack it so they still work or just remove them and make the breaking change.

When returning structs where one of the fields can be null use `Temperature?`

Reply:
Another option is to use: `Temperature temp = new Temperature(0, (UnitsNet.Units.TemperatureUnit)(-1));` - that is actually causing failure on any conversion or ToString:

```csharp
Temperature temp = new Temperature(0, (UnitsNet.Units.TemperatureUnit)(-1));
            Console.WriteLine(temp); // Throws NotImplementedException
            Console.WriteLine(temp.Unit); // Prints -1
            Console.WriteLine(temp.DegreesCelsius); // Throws NotImplementedException
```

another option is to create a single Invalid struct with reflection and then copy it in all places which need it